### DISCUSSION TITLE
Working on an extension.  Need to insert some javascript. No luck.

### DISCUSSION THREAD

**Discussion by enn-nafnlaus** [oobabooga/text-generation-webui#941] [General]
And I keep banging up against a wall on this one.  In ui(), I've tried all sorts of things.   gr.HTML('<script>alert("Hello, world!");</script>') , for example, never calls. I can see it in window.gradio_config when I view page source, but it never runs. I've tried appending into gr.shared["display"].value, no luck either.  No luck either with creating a gr.Interface.

How should I actually do this?

**Comment by poisenbery** [comment]
Bumping for similar issue:
Console indicates my custom extension has loaded, but it will not function at all.

**Reply by enn-nafnlaus** [reply]
> Bumping for similar issue: Console indicates my custom extension has loaded, but it will not function at all.

I can get output_modifier to function right.  Which means I technically could implement an awkward hack of simply injecting the javascript into the first chat message  ;)  But I assume we want something cleaner than that.

What functions are your extension overriding?

**Comment by nullgate** [comment]
IMHO, gradio is a nightmare for web developer.

**Comment by oobabooga** [comment]
The right way is using `_js` associated to some event. For instance, this is used to initialize some global javascript when the UI is launched

```python
shared.gradio['interface'].load(None, None, None, _js=f"() => {{{ui.main_js+ui.chat_js}}}")
```

where `ui.main_js` and `ui.chat_js` are strings containing javascript code.

In this case, the javascript function takes no inputs, but you can also pass interface parameters to the function like name1, context, temperature, etc.

**Reply by enn-nafnlaus** [reply]
> The right way is using `_js` associated to some event. For instance, this is used to initialize some global javascript when the UI is launched
> 
> ```python
> shared.gradio['interface'].load(None, None, None, _js=f"() => {{{ui.main_js+ui.chat_js}}}")
> ```
> 
> where `ui.main_js` and `ui.chat_js` are strings containing javascript code.
> 
> In this case, the javascript function takes no inputs, but you can also pass interface parameters to the function like name1, context, temperature, etc.

So should I hijack an event (such as load()) on an existing interface, or create a dummy interface to get its own events?  I don't want to mess up something else's scripting.  I guess I could just hijack anything that doesn't already have a .load function.

I don't actually need any js to run at the time of an event - I just need a function "rating_dialog" to be declared.  That way, the buttons added in output_modifier can function:

```
# Override output_modifier() function to insert html into chat strings
def output_modifier(chat_text):
  chat_html = ""
  chat_html += '<div class="rating-buttons">'
  chat_html += '<button class="like-button" onclick="rating_dialog(\'like\', \''+chat_text+'\')">👍</button>'
  chat_html += '<button class="dislike-button" onclick="rating_dialog(\'dislike\', \''+chat_text+'\')">👎</button>'
  chat_html += '</div>'
  chat_text += chat_html
  return chat_text

```
I have a framework setup (still needs debugging once the js function is inserted) so that, in javascript, it'll make a commit of the comment using the Github API to a repository of training data, including the chat text, like/dislike status, and the user's explanation of what they liked or disliked.  I imagine we'll accumulate training data very quickly if I can get it to work.

(Hmm, come to get it I should probably commit some sort of unique anonymized user ID as well, in case some user abuses the system so that they can be filtered out.... will think about it)

**Reply by oobabooga** [reply]
In this case, the best thing to do would be to add one more javascript string to the very same line that I gave as an example, like:

```python
custom_js = "..."
shared.gradio['interface'].load(None, None, None, _js=f"() => {{{ui.main_js+ui.chat_js+custom_js}}}")
```

If you get good results with this, I could implement user javascript extensions by pulling the `custom_js` from the `script.py` of an extension.

**Reply by enn-nafnlaus** [reply]
Unfortunately, I just can't get this approach to work.  Been messing with it endlessly the other day. Hopefully I'm not doing something stupid.

```
           extension_js = """
function test() { alert('Test called'); }
alert("This is a simple alert");
"""
            shared.gradio['interface'].load(None, None, None, _js=f"() => {{{ui.main_js+ui.chat_js+extension_js}}}")

```
When this code runs, the alert pops up, before the two errors at the bottom show up.

![image](https://user-images.githubusercontent.com/116288799/232116990-a5d06c22-a069-4f8e-a1de-cdb66520884d.png)

As soon as I close the alert, the first error shows up.  The second error shows up when I click a simple control with an onclick event that calls test().

Thoughts?  

If I can't find a way to do this cleanly, I guess I can take the really ugly tack of simply putting all of the code for the handler inside each onclick event... :Þ

**Comment by enn-nafnlaus** [comment]
ED: I think I'll just do this the ugly way and shove all the code into the onlick() function every single time so I can get it done. This isn't the 1990s here; nobody is on dialup and is going to care if a couple kilobytes get appended to every response from the AI.

**Comment by DavG25** [comment]
I had the same issue when making an extension and needed to inject JS.

I found that you can use the Gradio [`Blocks`](https://gradio.app/docs/#blocks) to create a new interface and inject JS when the UI loads, for example in your case you can do the following:

<br>

This would be the `script.py` file inside your extension's folder
```python
from pathlib import Path
import gradio as gr

# Initialize JS from file
with open(Path(__file__).resolve().parent / 'extension.js', 'r') as f:
    extension_js = f.read()

# Inject JS when Gradio loads the UI
def ui():
    # Use Gradio Blocks instead of shared.gradio['interface']
    # so we avoid conflicts with the current scripts (like ui.main_js and ui.chat_js)
    with gr.Blocks() as interface:
        interface.load(None, None, None, _js=f'() => {{{extension_js}}}')
```

And this would be the `extension.js` file inside your extension's folder
```javascript
// Display alert when JS loads
alert('Extension JS loaded');

// Define local functions as needed
function generateRatingText(action, text) {
  return `You rated the text '${text}' with a '${action}'`;
}

// To make functions accessible from the HTML DOM and browser console we use 'globalThis',
// this way the function can be accessed by the 'onclick' event in the HTML elements
globalThis.rating_dialog = function (action, text) {
  alert(generateRatingText(action, text));
}

```

<br>

Then you can add the `onclick` event as usual to call the functions defined with `globalThis`, for example this button
```html
<button onclick="rating_dialog('dislike', 'As an AI language model')">Dislike</button>
```
would call the global function `rating_dialog()` defined in the `extension.js` file and display the alert.

**Reply by enn-nafnlaus** [reply]
> I had the same issue when making an extension and needed to inject JS.
> 
> I found that you can use the Gradio [`Blocks`](https://gradio.app/docs/#blocks) to create a new interface and inject JS when the UI loads, for example in your case you can do the following:
> 
> 
> This would be the `script.py` file inside your extension's folder
> 
> ```python
> from pathlib import Path
> import gradio as gr
> 
> # Initialize JS from file
> with open(Path(__file__).resolve().parent / 'extension.js', 'r') as f:
>     extension_js = f.read()
> 
> # Inject JS when Gradio loads the UI
> def ui():
>     # Use Gradio Blocks instead of shared.gradio['interface']
>     # so we avoid conflicts with the current scripts (like ui.main_js and ui.chat_js)
>     with gr.Blocks() as interface:
>         interface.load(None, None, None, _js=f'() => {{{extension_js}}}')
> ```
> 
> And this would be the `extension.js` file inside your extension's folder
> 
> ```js
> // Display alert when JS loads
> alert('Extension JS loaded');
> 
> // Define local functions as needed
> function generateRatingText(action, text) {
>   return `You rated the text '${text}' with a '${action}'`;
> }
> 
> // To make functions accessible from the HTML DOM and browser console we use 'globalThis',
> // this way the function can be accessed by the 'onclick' event in the HTML elements
> globalThis.rating_dialog = function (action, text) {
>   alert(generateRatingText(action, text));
> }
> ```
> 
> 
> Then you can add the `onclick` event as usual to call the functions defined with `globalThis`, for example this button
> 
> ```
> <button onclick="rating_dialog('dislike', 'As an AI language model')">Dislike</button>
> ```
> 
> would call the global function `rating_dialog()` defined in the `extension.js` file and display the alert.

Sadly, this just doesn't work.

```
def ui():
  with gr.Blocks() as interface:
    extension_js = """
alert("Hello World #1!");    

function test() { alert('Hello world #2!'); }

alert("Hello World #3!");    
"""
    interface.load(None, None, None, _js=f'() => {{{extension_js}}}')
```

Both alerts #1 and #3 go off.  But test() still cannot be seen:

![image](https://user-images.githubusercontent.com/116288799/232318634-9aea80a0-3206-434e-8ad0-fb3eca4af9c2.png)

(The first javascript error occurs always, even if the extension isn't loaded)

Gradio seems to be some sort of parsing of the interface loading javascript, and it's apparently eating even trivial functions like test().  SIGH....

(I don't know how they managed to make something as trivial as adding a javascript function to a webapp this difficult, but they managed it...)

Oh, and also, this approach introduces a "loading" icon that never finishes.

Going back to an approach of trying to inject all code into every single chat box's onclick functions... though sadly, that has its own issues too, as now the app is parsing my HTML / javascript...

![image](https://user-images.githubusercontent.com/116288799/232319293-2d4f93d1-fbe2-41ac-9ec5-d352f953195b.png)

![image](https://user-images.githubusercontent.com/116288799/232319690-8ac410a9-590e-4266-bfd9-10a6c032e051.png)

Need to dig into why it's inserting quotation marks, and how to stop it... sigh...

(I expected my main hassles with this extension to be with GitHub, trying to find an easy way to make tokens that will let users commit but not alter existing content, and let me rollback / block vandalizing users.  But I can't even get to the point of even running my code to start trying to figure that out...)

**Reply by DavG25** [reply]
> Sadly, this just doesn't work.
> 
> ```
> def ui():
>   with gr.Blocks() as interface:
>     extension_js = """
> alert("Hello World #1!");    
> 
> function test() { alert('Hello world #2!'); }
> 
> alert("Hello World #3!");    
> """
>     interface.load(None, None, None, _js=f'() => {{{extension_js}}}')
> ```
> 
> Both alerts #1 and #3 go off. But test() still cannot be seen
> 
This is expected because the `test()` function gets defined in the context of the anonymous function that Gradio uses to load the JS.

What happens is this:

```javascript
// Global context
this; // 'this' is in the 'Window' context here, let's call it 'context 1'

// Gradio anonymous function that loads your JS
() => {
  this; // 'this' is in the context of the anonymous function, let's call it 'context 2'
  // Your JS code is executed here, in 'context 2'
}
```

In this case `test()` is defined in '_context 2_' and cannot be accessed from '_context 1_'.

If you want to access `test()` in '_context 1_' (so outside of the function where it gets defined), you need to define it using `globalThis`.

<br>

Does it work for you if you write the following instead?

```python
extension_js = """
alert("Hello World #1!");
globalThis.test = function () { alert('Hello world #2!'); }
alert("Hello World #3!");
"""

def ui():
  with gr.Blocks() as interface:
    interface.load(None, None, None, _js=f'() => {{{extension_js}}}')
```

**Reply by enn-nafnlaus** [reply]
Hey, success!

One issue remaining, though: as mentioned, adding in the gr.Blocks() creates this never-ending growing timer:

![image](https://user-images.githubusercontent.com/116288799/232326735-b389c096-61c9-4fb6-8b52-a0df487445f4.png)

Solutions?

**Reply by enn-nafnlaus** [reply]
Oh god...   is this going to be all the information I get for debugging?  An error but no reference to an actual line number in the Javascript that caused the error, just to a stacktrace of Gradio code?  :(  :(  :(

![image](https://user-images.githubusercontent.com/116288799/232327831-01eee8c3-d231-4e03-8665-6e83337f65c4.png)

**Reply by DavG25** [reply]
> Hey, success!
> 
> One issue remaining, though: as mentioned, adding in the gr.Blocks() creates this never-ending growing timer

That's strange, it doesn't seem to happen on my web UI, can you write the steps to replicate it or show the start-webui console output?

<br>

> Oh god... is this going to be all the information I get for debugging? An error but no reference to an actual line number in the Javascript that caused the error, just to a stacktrace of Gradio code? :( :( :(
> 
> ![image](https://user-images.githubusercontent.com/116288799/232327831-01eee8c3-d231-4e03-8665-6e83337f65c4.png)

You can do something like this to make debugging easier:
```python
import uuid
import gradio as gr

# The JS of the extension
extension_js = """
function test() {
  console.log('Executing test');
  throw 'Error';
}
test();
"""

# Just in case, to avoid conflicting HTML elements with the same ID
injector_uuid = uuid.uuid4();

# This JS will create a script tag in the DOM with the JS from
# 'extension_js' and execute it in the 'Window' context
js_injector = f"""
const deferredScript = document.getElementById('js-deferred-script-{injector_uuid}');
const script = document.createElement('script');
script.text = deferredScript.textContent;
document.head.appendChild(script);
deferredScript.remove();
"""

def ui():
    with gr.Blocks() as interface:
        # This HTML element is used to load the extension's JS in the 'Window'
        # context rather than the Gradio anonymous injection function
        gr.HTML(value=f'<deferred-script id="js-deferred-script-{injector_uuid}"> {extension_js} </deferred-script>', visible=False)
        # Load the injector's JS that will then load the extension's JS
        interface.load(None, None, None, _js=f'() => {{{js_injector}}}')
```
This will execute your `extension_js` code like it was in a normal `<script>` tag in HTML, so you can see which lines throw errors.

![1](https://user-images.githubusercontent.com/31524206/232334135-4080e1f1-a0fd-4c68-81e2-eede0f9a6318.png)

![2](https://user-images.githubusercontent.com/31524206/232334136-05e3a55e-3668-4b5d-976f-1a39d7d7220f.png)

At the same time, by doing this, the context of the `extension_js` script will change to '_Window_', so you won't need to use `globalThis` anymore.

**Reply by enn-nafnlaus** [reply]
Thanks.  :)  Unfortunately, your new version gives me:

![image](https://user-images.githubusercontent.com/116288799/232860073-382837ec-9f27-41a1-81c3-2568ce106000.png)

Reverting; I'll just go back to using the old "comment out code a couple dozen times until you find the problem" method.

As for the endless timer, here's my full code (excluding the large blocks of commented-out code, and the javascript file):

```
import io, re
from pathlib import Path

import gradio as gr
import modules.chat as chat
import modules.shared as shared

# Initialize JS from file
with open(Path(__file__).resolve().parent / 'extension.js', 'r') as f:
  extension_js = f.read()

# Inject JS when Gradio loads the UI
def ui():
  with gr.Blocks() as interface:
    interface.load(None, None, None, _js=f'() => {{{extension_js}}}')

# Override output_modifier() function to insert html into chat strings
input_buffer = ['', '', '']
output_buffer = ['', '', '']

def input_modifier(chat_text):
  global input_buffer
  input_buffer.append(chat_text.replace("'", "\\\'"))      # ATH: Use the internal history mechanism instead
  input_buffer = input_buffer[-3:]
  return chat_text

def output_modifier(chat_text):
  global input_buffer, output_buffer
  output_buffer.append(chat_text.replace("'", "\\\'"))   # ATH: Use the internal history mechanism instead
  output_buffer = output_buffer[-3:]
  chat_args = f'\'{input_buffer[0]}\', \'{output_buffer[0]}\', \'{input_buffer[1]}\', \'{output_buffer[1]}\', \'{input_buffer[2]}\', \'{output_buffer[2]}\''

```

As for the extension.js, I can strip it down to, say:

`alert("Hello World #1!"); `

Same problem.  Goes away when I strip the gr.Blocks() creation out.

**Reply by DavG25** [reply]
> Thanks. :) Unfortunately, your new version gives me:
> 
> ![image](https://user-images.githubusercontent.com/116288799/232860073-382837ec-9f27-41a1-81c3-2568ce106000.png)
> 
Yes that's the expected result! In the code I wrote previously, I included the line `throw 'Error';` that will generate an error in the console.
This was to demonstrate how you can make debugging easier, because clicking on "VM3143:4" on the right of the error will show you exactly where the code failed, instead of a stack trace.
You can disable this JS error by just removing that specific line.

If you're referring to the `TypeError` instead, it seems like it's a Gradio issue (see https://github.com/oobabooga/text-generation-webui/pull/1290) with Blocks when the `outputs` parameter is set to `None` and `_js` is used. 
You can fix that by adding a dummy output to the interface, like this:
```python
# Inject JS when Gradio loads the UI
def ui():
  with gr.Blocks() as interface:
    interface.load(None, None, gr.Label(visible=False), _js=f'() => {{{extension_js}}}')
```

<br><br>

> As for the endless timer, here's my full code (excluding the large blocks of commented-out code, and the javascript file):
> 
> ```
> import io, re
> from pathlib import Path
> 
> import gradio as gr
> import modules.chat as chat
> import modules.shared as shared
> 
> # Initialize JS from file
> with open(Path(__file__).resolve().parent / 'extension.js', 'r') as f:
>   extension_js = f.read()
> 
> # Inject JS when Gradio loads the UI
> def ui():
>   with gr.Blocks() as interface:
>     interface.load(None, None, None, _js=f'() => {{{extension_js}}}')
> 
> # Override output_modifier() function to insert html into chat strings
> input_buffer = ['', '', '']
> output_buffer = ['', '', '']
> 
> def input_modifier(chat_text):
>   global input_buffer
>   input_buffer.append(chat_text.replace("'", "\\\'"))      # ATH: Use the internal history mechanism instead
>   input_buffer = input_buffer[-3:]
>   return chat_text
> 
> def output_modifier(chat_text):
>   global input_buffer, output_buffer
>   output_buffer.append(chat_text.replace("'", "\\\'"))   # ATH: Use the internal history mechanism instead
>   output_buffer = output_buffer[-3:]
>   chat_args = f'\'{input_buffer[0]}\', \'{output_buffer[0]}\', \'{input_buffer[1]}\', \'{output_buffer[1]}\', \'{input_buffer[2]}\', \'{output_buffer[2]}\''
> ```

I'm not sure why it displays a loading timer for you, maybe you could try the implementation suggested by @ye7iaserag and see if that fixes the issue.

**Comment by ye7iaserag** [comment]
I have a very simple solution that works, but it doesn't look good
in your UI function add:
```
gr.HTML(value="<img src onerror='" + generate_js() + "'>")
```

now this function `generate_js` can return any js such as loading an entire js file and appending it to head:
```
def generate_js():
    js = f"""
    (() => {{
        let globalsScript = document.createElement("script");
        globalsScript.src = "file/extensions/{extension_name}/js/globals.js";
        document.head.appendChild(globalsScript);
    }})();
    """
    return js
```

Explanation:
Creating an `img` elemnt without a `src` will always trigger the `error` event, which will call the `onerror` listener.

**Reply by enn-nafnlaus** [reply]
Lol, amusing solution.  Will try it out tomorrow  :)
